code/game/turfs/baseturfs.dm

```diff
@@ -0,0 +1,188 @@
+			// Don't forget a lower index is lower in the baseturfs stack, the bottom is baseturfs[1]
+			target_baseturfs = copytarget.baseturfs.Copy(clamp(1 + ignore_bottom, 1 + copytarget.baseturfs.len - depth, copytarget.baseturfs.len))
+		else if(!ignore_bottom)
+			target_baseturfs = list(copytarget.baseturfs)
+		if(target_baseturfs)
+			target_baseturfs -= new_baseturfs & GLOB.blacklisted_automated_baseturfs
+			new_baseturfs += target_baseturfs
+
+	var/turf/newT = copytarget.copyTurf(src, copy_air)
+	newT.baseturfs = baseturfs_string_list(new_baseturfs, newT)
+	return newT
+
+/// Tries to find the given type in baseturfs.
+/// If found, returns how deep it is for use in other baseturf procs, or null if it cannot be found.
+/// For example, this number can be passed into ScrapeAway to scrape everything until that point.
+/turf/proc/depth_to_find_baseturf(baseturf_type)
+	var/index = baseturfs.Find(baseturf_type)
+	if (index == 0)
+		return null
+	return baseturfs.len - index + 1
+
+/// Returns the baseturf at the given depth.
+/// For example, baseturf_at_depth(1) will give the baseturf that would show up when scraping once.
+/turf/proc/baseturf_at_depth(index)
+	ASSERT(isnum(index))
+	if (islist(baseturfs))
+		return LAZYACCESS(baseturfs, baseturfs.len - index + 1)
+	else if (index == 1)
+		return baseturfs
+	else
+		return null
+
+/// Replaces all instances of needle_type in baseturfs with replacement_type
+/turf/proc/replace_baseturf(needle_type, replacement_type)
+	if (islist(baseturfs))
+		var/list/new_baseturfs
+
+		while (TRUE)
+			var/found_index = baseturfs.Find(needle_type)
+			if (found_index == 0)
+				break
+
+			new_baseturfs ||= baseturfs.Copy()
+			new_baseturfs[found_index] = replacement_type
+
+		if (!isnull(new_baseturfs))
+			baseturfs = baseturfs_string_list(new_baseturfs, src)
+	else if (baseturfs == needle_type)
+		baseturfs = replacement_type
+
+/// Removes all baseturfs that are found in the given typecache.
+/turf/proc/remove_baseturfs_from_typecache(list/typecache)
+	if (islist(baseturfs))
+		var/list/new_baseturfs
+
+		for (var/baseturf in baseturfs)
+			if (!typecache[baseturf])
+				continue
+
+			new_baseturfs ||= baseturfs.Copy()
+			new_baseturfs -= baseturf
+
+		if (!isnull(new_baseturfs))
+			baseturfs = baseturfs_string_list(new_baseturfs, src)
+	else if (typecache[baseturfs])
+		baseturfs = /turf/baseturf_bottom
+
+/// Returns the total number of baseturfs
+/turf/proc/count_baseturfs()
+	return islist(baseturfs) ? length(baseturfs) : 1
+
+/// Inserts a baseturf at the given level.
+/// "Level" here doesn't mean depth.
+/// For example, `insert_baseturf(2, /turf/open/floor/plating)` will make it so
+/// the 2nd to last turf in the list is plating.
+/// This is different from *depth*, since depth is the level from the top.
+/turf/proc/insert_baseturf(level, turf_type)
+	if (!islist(baseturfs))
+		assemble_baseturfs()
+		ASSERT(islist(baseturfs))
```

Comment:
similar concern 
also what does 
this replace? this seems weird I didn't think we stored baseturfs rolled up like this implies 
does not make sense in the context of shuttle code, its only use

Reply:
`insert_baseturf` only makes sense conceptually if you have a list of baseturfs. This is meant to guarantee that after `assemble_baseturfs()`, we are confident that we have a list, which is expected after this branch. I can replace this with `TEST_ONLY_ASSERT` if that is your concern.

Reply:
the idea with baseturfs is they get culled down to just a single path if there's only one of them. it's at the least valid
we don't expect to have baseturfs hidden inside other baseturfs definitions.
what you're doing here would pull new baseturfs from thin air. it doesn't make behavioral sense.
would be better to just add a case that adds the turf type either before or after the listified existing baseturf.

Reply:
```suggestion
		TEST_ONLY_ASSERT(islist(baseturfs))
```

Reply:
it shouldn't be test only, it doesn't make sense even in that context

Reply:
Explain what you mean then because this assertion is important for the code afterwards to not runtime

Reply:
I'm happy to change this back to ASSERT since I don't think it's hit during tests but the expectation that baseturfs is a list holds whether this line exists or not